src/WebJobs.Script/Workers/FunctionDataCache/FunctionDataCache.cs

```diff
@@ -0,0 +1,329 @@
+                    }
+
+                    long bytesRequired = sharedMemoryMeta.Count;
+                    if (!EvictUntilCapacityAvailable(bytesRequired))
+                    {
+                        return false;
+                    }
+
+                    // Add the mapping into the local cache
+                    _localCache.Add(cacheKey, sharedMemoryMeta);
+
+                    // Update the LRU list (mark this key as the most recently used)
+                    AddToEndOfLRU(cacheKey);
+
+                    if (isIncrementActiveReference)
+                    {
+                        IncrementActiveReference(cacheKey);
+                    }
+
+                    // Update the cache utilization
+                    RemainingCapacityBytes -= sharedMemoryMeta.Count;
+
+                    success = true;
+                    return true;
+                }
+            }
+            finally
+            {
+                if (!success && isDeleteOnFailure)
+                {
+                    if (!_sharedMemoryManager.TryFreeSharedMemoryMap(sharedMemoryMeta.MemoryMapName))
+                    {
+                        _logger.LogTrace("Cannot free shared memory map: {MapName} with size: {Size} bytes on failure to insert into the cache", sharedMemoryMeta.MemoryMapName, sharedMemoryMeta.Count);
+                    }
+                }
+            }
+        }
+
+        public bool TryGet(FunctionDataCacheKey cacheKey, bool isIncrementActiveReference, out SharedMemoryMetadata sharedMemoryMeta)
+        {
+            lock (_lock)
+            {
+                // Get the value from the local cache
+                if (!_localCache.TryGetValue(cacheKey, out sharedMemoryMeta))
+                {
+                    // Key does not exist in the local cache
+                    return false;
+                }
+
+                // Update the LRU list (mark this key as the most recently used)
+                AddToEndOfLRU(cacheKey);
+
+                if (isIncrementActiveReference)
+                {
+                    IncrementActiveReference(cacheKey);
+                }
+
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Note: This will remove the entry even if it has active references. It is the responsibility of the caller to
+        /// ensure the entry is safe to be removed.
+        /// </summary>
+        /// <param name="cacheKey">The <see cref="FunctionDataCacheKey"/> corresponding to the entry to be removed.</param>
+        /// <returns><see cref="true"/> if the entry was successfully removed, <see cref="false"/> if not.</returns>
+        public bool TryRemove(FunctionDataCacheKey cacheKey)
+        {
+            lock (_lock)
+            {
+                // TODO remove this Get and check if we can do remove with *out*
+                if (!TryGet(cacheKey, isIncrementActiveReference: false, out SharedMemoryMetadata sharedMemoryMeta))
```

Comment:
This is a re-entrant lock, and since it's on the same object in both methods, it won't deadlock. 
However, for readability I can see if wrapping it into private methods is easier to understand.

Reply:
Ah you are right! I didn't know that the `lock` block makes it re-entrant. Thanks! That clears it.